File: gestion-ecole/src/main/java/ma/edu/gestionecole/gestionecole/entities/Etudiant.java
package ma.edu.gestionecole.gestionecole.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.springframework.format.annotation.DateTimeFormat;

import javax.persistence.*;
import javax.validation.constraints.Email;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.util.Date;
import java.util.List;

@Entity
@Data @AllArgsConstructor @NoArgsConstructor @ToString
public class Etudiant
{
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @NotBlank(message = "Le Nom est Obligatoire") @Size(max = 30)
    private String nom;
    @NotBlank(message = "Le Prenom est Obligatoire") @Size(max = 30)
    private String prenom;
    @NotNull @Size(max = 1) // M / F
    private String genre;
    @Temporal(TemporalType.DATE)
    @DateTimeFormat(pattern = "yyyy-MM-dd")
    private Date dateNaissance;
    private String adresse;
    @Email
    //unique
    private String email;

    @OneToMany(mappedBy = "etudiant", cascade = CascadeType.ALL)
    @JsonIgnore
    private List<Note> notes;

    @OneToMany(mappedBy = "etudiant", cascade = CascadeType.ALL)
    @JsonIgnore
    private List<Absence> absences;

    @ManyToOne
    @NotNull
    private Classe classe;
}

File: ecole-frontend/src/app/etudiant.ts
import {Classe} from "./classe";

export class Etudiant {
  id: number = 0;
  nom: string = '';
  prenom: string = '';
  genre: string = '';
  dateNaissance!: Date; //the '!' means that 'dateNaissance' will be assigned a value before it's accessed, even if it's not initialized in the constructor.
  adresse: string = '';
  email: string = '';
  classe: Classe = new Classe();
//  notes: Note[];
//  absences: Absence[];


  constructor() {
  }

/*  constructor(
    id: number,
    nom: string,
    prenom: string,
    genre: string,
    dateNaissance: Date,
    adresse: string,
    email: string,
//    notes: Note[],
//    absences: Absence[],
//    classe: Classe
  ) {
    this.id = id;
    this.nom = nom;
    this.prenom = prenom;
    this.genre = genre;
    this.dateNaissance = dateNaissance;
    this.adresse = adresse;
    this.email = email;
//    this.notes = notes;
//    this.absences = absences;
//    this.classe = classe;
  }*/
}
